chain/network/src/rate_limits/messages_limits.rs

```diff
@@ -0,0 +1,320 @@
+    }
+}
+
+/// Rate limit configuration for a single network message.
+#[derive(Clone)]
+pub struct SingleMessageConfig {
+    pub message_key: RateLimitedPeerMessageKey,
+    pub maximum_size: u32,
+    pub refill_rate: f32,
+    /// Optional initial size. Defaults to `maximum_size` if absent.
+    pub initial_size: Option<u32>,
+}
+
+impl SingleMessageConfig {
+    pub fn new(
+        message_key: RateLimitedPeerMessageKey,
+        maximum_size: u32,
+        refill_rate: f32,
+        initial_size: Option<u32>,
+    ) -> Self {
+        Self { message_key, maximum_size, refill_rate, initial_size }
+    }
+}
+
+/// Network messages rate limits configuration.
+#[derive(Default, Clone)]
+pub struct Config {
+    pub rate_limits: Vec<SingleMessageConfig>,
+}
+
+impl Config {
+    /// Validates this configuration object.
+    ///
+    /// # Errors
+    ///
+    /// If at least one error is present, returns the list of all configuration errors.  
+    pub fn validate(&self) -> Result<(), Vec<(RateLimitedPeerMessageKey, TokenBucketError)>> {
+        let mut errors = Vec::new();
+        for message_config in &self.rate_limits {
+            if let Err(err) = TokenBucket::validate_refill_rate(message_config.refill_rate) {
+                errors.push((message_config.message_key, err));
+            }
+        }
+        if errors.is_empty() {
+            Ok(())
+        } else {
+            Err(errors)
+        }
+    }
+}
+
+/// This enum represents the variants of [PeerMessage] that can be rate limited.
+/// It is meant to be used as an index for mapping peer messages to a value.
+#[derive(Clone, Copy, enum_map::Enum, strum::Display, Debug, PartialEq, Eq)]
+#[allow(clippy::large_enum_variant)]
+pub enum RateLimitedPeerMessageKey {
+    SyncRoutingTable,
+    DistanceVector,
+    RequestUpdateNonce,
+    SyncAccountsData,
+    PeersRequest,
+    PeersResponse,
+    BlockHeadersRequest,
+    BlockHeaders,
+    BlockRequest,
+    Block,
+    Transaction,
+    SyncSnapshotHosts,
+    StateRequestHeader,
+    StateRequestPart,
+    VersionedStateResponse,
+    BlockApproval,
+    ForwardTx,
+    TxStatusRequest,
+    TxStatusResponse,
+    StateResponse,
+    PartialEncodedChunkRequest,
+    PartialEncodedChunkResponse,
+    VersionedPartialEncodedChunk,
+    PartialEncodedChunkForward,
+    ChunkEndorsement,
+    ChunkStateWitnessAck,
+    PartialEncodedStateWitness,
+    PartialEncodedStateWitnessForward,
+}
+
+/// Given a `PeerMessage` returns a tuple containing the `RateLimitedPeerMessageKey`
+/// corresponding to the message's type and its the cost (in tokens) for rate limiting
+/// purposes.
+///
+/// Returns `Some` if the message has the potential to be rate limited (through the correct configuration).
+/// Returns `None` if the message is not meant to be rate limited in any scenario.
+fn get_key_and_token_cost(message: &PeerMessage) -> Option<(RateLimitedPeerMessageKey, u32)> {
+    use RateLimitedPeerMessageKey::*;
+    match message {
```

Comment:
More boilerplate than I would like, but it make sure we don't forget to consider rate limits when adding new messages.